pkg/controller/volume/selinuxwarning/cache/volumecache.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"sort"
+	"sync"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// Log level at which the volume cache will be dumped after each change.
+	dumpLogLevel = 10
+)
+
+type VolumeCache interface {
+	// Add a single volume to the cache. Returns list of conflicts it caused.
+	AddVolume(logger klog.Logger, volumeName v1.UniqueVolumeName, podKey cache.ObjectName, label string, changePolicy v1.PodSELinuxChangePolicy, csiDriver string) []Conflict
+
+	// Remove a pod from the cache. Prunes all empty structures.
+	DeletePod(logger klog.Logger, podKey cache.ObjectName)
+
+	// GetPodsForCSIDriver returns all pods that use volumes with the given CSI driver.
+	GetPodsForCSIDriver(driverName string) []cache.ObjectName
+
+	// SendConflicts sends all current conflicts to the given channel.
+	SendConflicts(logger klog.Logger, ch chan<- Conflict)
+}
+
+// VolumeCache stores all volumes used by Pods and their properties that the controller needs to track,
+// like SELinux labels and SELinuxChangePolicies.
+type volumeCache struct {
+	mutex sync.Mutex
+	// All volumes of all existing Pods.
+	volumes map[v1.UniqueVolumeName]usedVolume
+}
+
+var _ VolumeCache = &volumeCache{}
+
+// NewVolumeLabelCache creates a new VolumeCache.
+func NewVolumeLabelCache() VolumeCache {
+	return &volumeCache{
+		volumes: make(map[v1.UniqueVolumeName]usedVolume),
+	}
+}
+
+// usedVolume is a volume that is used by one or more existing pods.
+// It stores information about these pods to detect conflicts and generate events.
+type usedVolume struct {
+	csiDriver string
+	// List of pods that use this volume. Indexed by pod key for faster deletion.
+	pods map[cache.ObjectName]podInfo
+}
+
+// Information about a Pod that uses a volume.
+type podInfo struct {
+	// SELinux label to be applied to the volume in the Pod.
+	// Either as mount option or recursively by the container runtime.
+	label string
```

Comment:
same about renaming `label` to `selinuxLabel`

Reply:
Renamed